### PR TITLE
llama: map each contiguous run of a context's tensors, not one span over all of them

### PR DESCRIPTION
**Pull request by danielhanchen**
`get_mapping_range` returns a single `[first, last)` computed as the min and max offset of the tensors a context owns in one file. A min/max cannot express a gap, so when a large tensor belonging to a **different** buffer type sits between them, that tensor is inside the mapped range.

On a backend that copies tensor by tensor this costs nothing, which is why it has gone unnoticed. On a backend that maps the host pointer it does not: `buffer_from_host_ptr` is called on the whole range, and everything in the gap becomes resident for the life of the process.

Four backends advertise the capability. CPU, BLAS and zendnn only wrap the pointer, so the range costs address space and nothing else. Metal creates real `MTLBuffer` objects over it with `newBufferWithBytesNoCopy`, counted against `recommendedMaxWorkingSetSize`, and pins them in a residency set on macOS 15 and later.

## Where it bites

Qwen3.8-Flash-Next places `per_layer_token_embd`, 26.8 GiB and CPU resident, at tensor index 4 of 373 in its second shard. Every other tensor in that shard belongs to the GPU context, on both sides of it.

Reproducible on Linux, because the CPU backend takes the same path and the PLE table lives in its own lazy context. `-ngl 0`, UD-IQ1_S, and llama.cpp's own per buffer log:

| | `CPU_Mapped` buffers | total mapped |
|---|---|---|
| before | 47674.92 + 21500.26 + 27465.95 | **96641.13 MiB** |
| after | 347.70 + 19861.28 + 21500.26 + 27465.95 | **69175.19 MiB** |

The shard 2 span splits into 347.70 + 19861.28 = 20208.98 MiB, and 47674.92 - 20208.98 = 27465.94 MiB, the PLE table to the byte. The model file is 69165 MiB, so the old span over maps it by 39.7 percent and the new one lands on it.

## What the change does

Map each contiguous run of a context's tensors instead of one span over all of them.

Runs closer together than 32 MiB are merged, so a model whose tensors are laid out per buffer type still produces exactly one buffer and one mapping, and only a genuinely large foreign tensor splits the range. If some layout yields more than 64 runs, it falls back to the single span rather than create a buffer per tensor.

`ggml_backend_tensor_alloc` already asserts that the address lies inside the buffer it is given, so a tensor resolved against the wrong run aborts rather than mapping silently to the wrong place.

## Testing

Built at `9723942ad`, CUDA, B200.

- Byte identical greedy output against the unpatched base on Qwen3.8-Flash-Next UD-IQ1_S, Qwen3.8-27B Q4_K_M and Qwen3-0.6B Q4_K_M. The harness runs the base against itself first and refuses to report anything if that is not reproducible.
- Byte identical image captioning through `llama-mtmd-cli`, which is the only path that exercises the M-RoPE branch.
- `test-llama-archs` for qwen4exp, llama, qwen3next, gemma3n and qwen3moe, checked for real result rows rather than exit code alone.
- `git diff --stat -- ggml/` is empty, so `test-backend-ops` cannot say anything this diff does not already guarantee.

## What is not tested here

There is no Apple hardware on this box. The Metal consequence is derived from the source plus the byte accounting above, and cross checked against the memory figures published for a third party GGUF of the same model that happens to isolate the table in its own shard. The buffer counts and sizes in the table are measured; the Metal residency saving is not.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Refreshed onto `b10786`, which is the base tonight's nightly will pick. The PR base branch moved to `base/upstream-de8656bd9`, which is that tag verbatim, so the diff here is still this feature and nothing else.

What the move required is in the merge commit message. Verified with the whole 13-pin set replayed on `b10786`: every pin merges, `pin_contract.py` reports all 13 intact, the CUDA `sm_100` build is clean, and the feature matrix passes all six features on a B200. The full output is on #174.

**Comment by danielhanchen** [comment]
The `CI (webgpu)` failure on this branch does not reproduce, so I am not treating it as a bug in this PR.

What it showed was one row of `test-llama-archs`:

```
| qwen2moe|WebGPU|MoE|  OK     (-nan)|     FAIL|
```

`-nan` means the logits were degenerate, and the roundtrip then compares NaN against itself and reports a difference. Two follow-up runs disagree with it:

| ref | qwen2moe WebGPU | runner CPU |
|---|---|---|
| `base/upstream-de8656bd9` (b10786, no pins) | `OK (1.15e-06)`, roundtrip OK | EPYC 7763 |
| this branch, re-run | `OK (1.09e-06)`, roundtrip OK | EPYC 9V74 |
| this branch, original | `-nan`, roundtrip FAIL | Xeon Platinum 8573C |

The base is clean and this branch is clean on a second run, so the first result was noise. The one correlation worth recording, in case it recurs: the failing run is the only one of the three that landed on an Intel runner, and the NaN appeared on the WebGPU and Meta rows while the CPU reference row was exact. That is a hint about the reference, not proof of anything.

Reading the diff supports the same conclusion. The per-run buffer lookup selects the buffer whose address range contains the tensor, which is correct, and `test-llama-archs` builds its models from an in-memory gguf with no file and no mmap, so the run-splitting path this PR adds is not even reached by that test.

**Comment by Blackskyliner** [comment]
> There is no Apple hardware on this box. The Metal consequence is derived from the source plus the byte accounting above, and cross checked against the memory figures published for a third party GGUF of the same model that happens to isolate the table in its own shard. The buffer counts and sizes in the table are measured; the Metal residency saving is not.

I could provide the testing on a MBP 128G Apple M3 Max if you give me a runbook which generates the needed output for verification.
